Ska/engarchive/update_client_archive.py

```diff
@@ -125,22 +133,225 @@ class DelayedKeyboardInterrupt(object):
         signal.signal(signal.SIGINT, self.old_handler)
         if self.signal_received:
             print('Shutting down due to keyboard interrupt', file=sys.stderr)
             sys.exit(1)
 
 
+def add_msids_from_file(opt, logger):
+    """
+    Main function for adding MSIDs to the local archive from a remote server or disk.
+
+    :param opt: options
+    :param logger: logger
+    :return: None
+    """
+    msids, msids_content = get_msids_for_add_msids_from_file(opt, logger)
+    copy_files = get_copy_files(logger, msids, msids_content)
+    if not copy_files:
+        logger.info(f'Local cheta archive is complete for MSIDs in {opt.add_msids_from_file}')
+        return
+
+    if '@' in opt.server_data_root:
+        copy_server_files_ssh(opt, logger, copy_files)
+    else:
+        copy_server_files(opt, logger, copy_files, opt.server_data_root, copy_func=shutil.copyfile)
+
+
+def copy_server_files(opt, logger, copy_files, server_path, copy_func, as_posix=False):
+    """
+    Copy list of files from server to local, making directories as needed.
+
+    This takes a function copy(src, dest) that could be a local copy or could be
+    a network copy.
+
+    :param opt: options
+    :param logger:  logger
+    :param copy_files: list of file paths, relative to the cheta archive root
+    :param server_path: path representing the cheta archive root on the server
+    :param copy_func: function to copy(src, dest)
+    :return: None
+    """
+    from astropy.utils.console import ProgressBar
+
+    logger.info(f'Copying {len(copy_files)} files from {opt.server_data_root} to {opt.data_root}')
+    if opt.dry_run:
+        logger.info('DRY RUN')
+
+    with ProgressBar(len(copy_files)) as progress_bar:
+        for copy_file in copy_files:
+            progress_bar.update()
+            local_file = Path(opt.data_root, copy_file)
+            server_file = Path(server_path, copy_file)
+            if as_posix:
+                server_file = server_file.as_posix()
+            local_file.parent.mkdir(parents=True, exist_ok=True)
+            if not opt.dry_run:
+                with DelayedKeyboardInterrupt(logger):
+                    logger.verbose('\ncopy server:{} local:{}'
+                                   .format(str(server_file), str(local_file)))
+                    copy_func(str(server_file), str(local_file))
+
+
+def copy_server_files_ssh(opt, logger, copy_files):
+    """
+    Copy list of files from a remote server to the local archive.
+
+    This uses options in ``opt`` to deterimine the user name, remote server name,
+    and remote server path.  It then connects to the server, makes an sftp calls
+    ``copy_server_files`` to do the copy.
+
+    :param opt: options
+    :param logger: logger
+    :param copy_files: list of file paths, relative to the cheta archive root
+    :return: None
+    """
+    import paramiko
+
+    match = re.match(r'(\w+)@([\w.]+)(:.+)?', opt.server_data_root)
```

Comment:
Of course.  If the `.` retained  the "any-character" meaning in a set, it would be no different than just using a dot.  It looks like this is fairly standard (even in MATLAB).  Thanks for setting me straight.